### ISSUE TITLE
Discussion: URL Generation for "Magic Links"

### ISSUE THREAD

**Issue by MisterJames** [HTBox/allReady#512]
This is a breakout from #505 and related functionality where we want a user to respond to a message or notification.

We should have a separate function call/library that we use that can create a link, the result of which should be able to log in a specific user and either take them to a specific page or execute a specific action.

Examples might be:
 - confirming task assignment
 - confirming activity changes
 - approving/rejecting changes
 - unfollowing a campaign or activity

The objective here would be that a user can react to an SMS or email message without having to sign into the site. This is similar to mobile functionality for apps like Slack (magic sign in link) or MailChimp (confirm subscription) where a user does not need to sign in.

The link should expire after a set amount of time (perhaps default to 24 hours).

The system should periodically run a job to check for (and delete) expired links.

@BillWagner @dpaquette @mheggeseth @tonysurma Hoping for your thoughts/input!

**Comment by mmoore99** [comment]
@MisterJames  [here](http://stackoverflow.com/a/33873988/2645) is a link to an interesting approach to generating a token to be used in "Magic Links".  In this approach the the user id, guid, expiration, reason code are combined together and encrypted so that when the endpoint receives the code it can be validated without the need for having to persist the code anywhere. I guess one of the downsides to this approach is that the code could be used multiple times within its valid time frame.  However since these links would typically be targeted to a specific endpoint to execute some action, the endpoint would be able to determine if the action had already been taken.

**Comment by MisterJames** [comment]
@mmoore99 That approach could easily be modified to prevent duplicate action if all that was persisted was the token, after the fact. The first thing the "resolver" would do, after decrypting the payload would be to check if the token's been executed. Again, tokens are only valid for so long, so if you persist the expiry date of the token, a job could execute daily to clean up any older pieces. All pretty cache-able, all pretty straightforward.

**Comment by stevejgordon** [comment]
@MisterJames @mmoore99 I've been digging into the source of Asp.Net Core Identity for some blog posts I'm working on at the moment. Identity already has the concept of a token provider IUserTokenProvider. An implementation of one can be seen at https://github.com/aspnet/Identity/blob/dev/src/Microsoft.AspNetCore.Identity/EmailTokenProvider.cs
 and https://github.com/aspnet/Identity/blob/dev/src/Microsoft.AspNetCore.Identity/TotpSecurityStampBasedTokenProvider.cs

This already looks to do much of what is being described above. This particular provider uses the SecurityStamp field in the User table as part of the code generation. Identity provides methods to update that security stamp, which would then invalidate any codes issued against the old SecurityStamp. In the cases above we may need to persist different security stamps for different tasks / sets of tasks.

It seems to me that this provider or a custom implementation could very well handle our requirements?

I'm still reading and stepping through the Identity and DataProtection code to understand exactly how it all works, but it doesn't seem too complex. I'd be interested to put into practice some of what I'm writing about and contribute to this in some way.

**Comment by dpaquette** [comment]
I really like this idea. It would avoid the user getting lost or confused by having to log in after clicking a link. 

My initial thought was that we would us some sort of token provider like @stevejgordon has suggested. We definitely want to the aspnetcore framework bits wherever we can for this since it will be much more secure than anything we could build on our own.

**Comment by MisterJames** [comment]
Agreed, @stevejgordon and @dpaquette if we can use bits that are already in place (or derive from them) then we have less heavy lifting to do and run less risk of opening a hole we don't want.  I think we should explore this.

**Comment by stevejgordon** [comment]
I've finally finished my post looking at token providers (hopefully accurate in my conclusions).

http://stevejgordon.co.uk/asp-net-core-identity-token-providers

A simple token provider would probably meet our needs here. It would probably be good to start with a list of all cases where we would want to send a link to verify an action. Once we know what cases we need to generate tokens and links for we just need to decide if the issue of a new token would need to invalidate the old one. In most of the cases above it feels like we simply need to provide a link (valid for x hours) which could confirm a specific action for a specific activity / task etc.

I doubt we would need to persist much (if anything) in those cases as the encrypted token could probably hold the information necessary).

In the case of SMS are you thinking we send a code which the user then verifies somewhere? I'm not sure a link would work too well as it could be quite long.

**Comment by tonysurma** [comment]
@MisterJames @dpaquette @stevejgordon any outcome of this?
